src/Microsoft.Health.Fhir.Api/Controllers/AadProxyController.cs

```diff
@@ -0,0 +1,355 @@
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net.Http;
+using System.Text;
+using System.Threading.Tasks;
+using EnsureThat;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using Microsoft.Health.Fhir.Core.Configs;
+using Microsoft.Health.Fhir.Core.Exceptions;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.Health.Fhir.Api.Controllers
+{
+    /// <summary>
+    /// Controller class enabling Azure Active Directory SMART on FHIR Proxy Capability
+    /// </summary>
+    [Route("/AadProxy")]
+    public class AadProxyController : Controller
+    {
+        private readonly SecurityConfiguration _securityConfiguration;
+        private readonly bool _isAadV2;
+        private readonly ILogger<SecurityConfiguration> _logger;
+        private readonly IHttpClientFactory _httpClientFactory;
+        private readonly string _aadAuthorizeEndpoint;
+        private readonly string _aadTokenEndpoint;
+
+        // TODO: _launchContextFields contain a list of fields that we will transmit as part of launch context, should be configurable
+        private readonly string[] _launchContextFields = { "patient", "encounter", "practitioner", "need_patient_banner", "smart_style_url" };
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="AadProxyController" /> class.
+        /// </summary>
+        /// <param name="securityConfiguration">Security configuration parameters.</param>
+        /// <param name="httpClientFactory">HTTP Client Factory.</param>
+        /// <param name="logger">The logger.</param>
+        public AadProxyController(IOptions<SecurityConfiguration> securityConfiguration, IHttpClientFactory httpClientFactory, ILogger<SecurityConfiguration> logger)
+        {
+            EnsureArg.IsNotNull(securityConfiguration, nameof(securityConfiguration));
+
+            _securityConfiguration = securityConfiguration.Value;
+            _isAadV2 = new Uri(_securityConfiguration.Authentication.Authority).Segments.Contains("v2.0");
+            _logger = logger;
+            _httpClientFactory = httpClientFactory;
+
+            var openIdConfigurationUrl = $"{_securityConfiguration.Authentication.Authority}/.well-known/openid-configuration";
+
+            HttpResponseMessage openIdConfigurationResponse;
+            using (var httpClient = httpClientFactory.CreateClient())
+            {
+                try
+                {
+                    openIdConfigurationResponse = httpClient.GetAsync(new Uri(openIdConfigurationUrl)).GetAwaiter().GetResult();
+                }
+                catch (Exception ex)
```

Comment:
We should avoid catching all exceptions types.

Reply:
I adopted this pattern from `CapabilityStatementExtensions.cs`. There could be a few different things to catch. Do you propose we walk through all of them?

Reply:
HttpRequestException and OperationCanceledException should cover it. Also, I'd just do `openIdConfigurationResponse.EnsureSuccessStatusCode()` instead of the if block below.